warp10/src/main/java/io/warp10/script/functions/MSTORE.java

```diff
@@ -35,42 +36,66 @@ public MSTORE(String name, boolean conditional) {
   public MSTORE(String name) {
     this(name, false);
   }
-  
+
   @Override
   public Object apply(WarpScriptStack stack) throws WarpScriptException {
+    List optionnalKeyList = null;
     Object o = stack.pop();
 
+    if (o instanceof List) {
+      optionnalKeyList = (List) o;
+      o = stack.pop();
+    }
     if (!(o instanceof Map)) {
       throw new WarpScriptException(getName() + " expects a Map as argument.");
     }
 
     Map<Object, Object> variables = (Map<Object, Object>) o;
 
     // Check that each key of the map is either a LONG or a STRING
-    for (Object elt: variables.keySet()) {
-      if (null != elt && (!(elt instanceof String) && !(elt instanceof Long))) {
-        throw new WarpScriptException(getName() + " expects each key to be a STRING or a LONG.");
-      }
-    }
-
-    for (Map.Entry<Object, Object> entry: variables.entrySet()) {
-      Object symbol = entry.getKey();
+    if (null != optionnalKeyList) {
+      // Check that each element of the list is either a LONG or a STRING
+      for (Object symbol: optionnalKeyList) {
+        if (null != symbol && (!(symbol instanceof String) && !(symbol instanceof Long))) {
+          throw new WarpScriptException(getName() + " expects a list of variable names or register numbers as second argument.");
+        }
+        Object v = variables.get(symbol);
 
-      if (null == symbol) {
-        continue;
+        if (symbol instanceof Long) {
+          if (!conditional || null == stack.load(((Long) symbol).intValue())) {
+            stack.store(((Long) symbol).intValue(), v);
+          }
+        } else {
+          if (!conditional || !stack.getSymbolTable().containsKey((String) symbol)) {
+            stack.store((String) symbol, v);
+          }
+        }
+      }
+    } else {
+      for (Object elt: variables.keySet()) {
```

Comment:
Simply check here that elt is in optionalKeyList or not, so as to reuse code for better readability.

Reply:
If the input map has lots of keys, iteration on the optional list is more efficient.  (I don't say it is a good use of this function... but it may happen)

Reply:
I don't think we would see any difference by optimizing looping through maps in comparison than loop done through GTS data for example.
Anyway, it is still possible to refactor the code using one iterator while loop rather than two for loops:
```
Iterator itr = null != optionalKeyList ? optionalKeyList.iterator() : variables.keySet().iterator();
    while(itr.hasNext()) {  ...  }
```